File: JavaAndScala/src/main/scala/org/lang/scala/conclusion/design/factory/TestFactory.scala
package org.lang.scala.conclusion.design.factory

import org.lang.scala.common.utils.ReflectionUtils
import org.lang.scala.conclusion.design.factory.abstracts.BrandFactory
import org.lang.scala.conclusion.design.factory.abstracts.MacFactory
import org.lang.scala.conclusion.design.factory.abstracts.WindowsFactory
import org.lang.scala.conclusion.design.factory.configuration.TxtConfigurationFactory
import org.lang.scala.conclusion.design.factory.configuration.common.Clerk
import org.lang.scala.conclusion.design.factory.configuration.common.Customer
import org.lang.scala.conclusion.design.factory.method.ConfigurationFactory
import org.lang.scala.conclusion.design.factory.method.PropConfigurationFactory
import org.lang.scala.conclusion.design.factory.method.XmlConfigurationFactory
import org.lang.scala.conclusion.design.factory.method.common.Lieutenant
import org.lang.scala.conclusion.design.factory.method.common.Sergeant
import org.lang.scala.conclusion.design.factory.simple.LaptopFactory
import org.lang.scala.conclusion.design.factory.simple.VehicleFactory
import org.lang.scala.conclusion.design.factory.singleton.design.LazySingletonFactory
import org.lang.scala.conclusion.design.factory.singleton.design.common.Teacher
import org.lang.scala.conclusion.design.factory.singleton.objects.Professor

/**
 * 	This is a stand-alone object to test factory pattern
 * 
 * 	There are several types of factory pattern, of which those commonly-used including:
 *  -- Product-level: Factory for the product
 *     -- Simple factory: to return an instance from different class according to different arguments
 *     -- Singleton factory: to return a singleton instance of any class
 *     -- Configuration factory: to return an instance from configuration file
 *        -- Typically applied in the design of Java frameworks (e.g., Spring, etc.)
 *  -- Factory-level: Based on low-level with factory abstraction
 *     -- Factory method: abstract the factory and delegates the instance creation to its child classes
 *        -- The instance (or product created by factory) belongs to one family
 *     -- Abstract factory: based on factory method and return a multiple families of instance
 *        -- The instance (or product created by factory) belongs to multiple families
 *  
 *  Factory pattern is usually implemented with reflection
 * 
 * 	@author VinceYuan
 */
object TestFactory {
 
  /**
   * 	This is a main method for execution
   */
  def main(args: Array[String]): Unit = {
    println("Here tests simple factory:")
    testSimpleFactory()
    println("\nHere tests singleton factory:")
    testSingletonFactory()
    println("\nHere tests configuration factory:")
    testConfigurationFactory()  
    println("\nHere tests factory method:")
    testFactoryMethod()         
    println("\nHere tests abstract factory:")
    testAbstractFactory()
  }
  
  /**
   * 	This is a method to test simple factory
   */
  private def testSimpleFactory(): Unit = {
    
		/*	Get different instances of the parent interface according to the argument	*/
		val acer = LaptopFactory.getLaptop("Acer")
		val dell = LaptopFactory.getLaptop("Dell")
		val lenovo = LaptopFactory.getLaptop("Lenovo")
		acer.startup()
		dell.startup()
		lenovo.startup()
		
		/*	Get different instances of the parent abstract class according to the argument	*/
		val ship = VehicleFactory.getVehicle("Ship")
		val train = VehicleFactory.getVehicle("Train")
		val truck = VehicleFactory.getVehicle("Truck")
		ship.turn()
		train.turn()
		truck.turn()
  }
  
  /**
   * 	This is a method to test singleton factory
   */
  private def testSingletonFactory(): Unit = {
    
    /*	Here tests singleton factory through design pattern for Scala only	*/
    var vince = LazySingletonFactory.getInstance[Teacher]
    var violet = LazySingletonFactory.getInstance[Teacher]
    println(s"Vince: ${vince}")
    println(s"Violet: ${violet}")
    println(s"If vince and violet reference the same instance? ${vince == violet && vince.equals(violet)}")
    
    /*	Here tests singleton factory through design pattern for Java compatibility	 */
    val clazz = ReflectionUtils.getJavaClassFromScalaType[Teacher]
    vince = LazySingletonFactory.getInstance(clazz).asInstanceOf[Teacher]
    violet = LazySingletonFactory.getInstance(clazz).asInstanceOf[Teacher]
    println(s"Vince: ${vince}")
    println(s"Violet: ${violet}")
    println(s"If vince and violet reference the same instance? ${vince == violet && vince.equals(violet)}")
    
    /*	Here tests singleton factory through object	directly	*/
    val tom = Professor
    val jerry = Professor
    println(s"If tom and jerry reference the same instance? ${tom == jerry && tom.equals(jerry)}")
  }
  
  /**
   * 	This is a method to test configuration factory
   */
  private def testConfigurationFactory(): Unit = {
    
    /*	Get instances from configuration files	*/
    val clerkByJava = TxtConfigurationFactory.getInstanceByJavaReflection("clerk_config").asInstanceOf[Clerk]
    val clerkByScala = TxtConfigurationFactory.getInstanceByScalaReflection[Clerk]("clerk_config")
    val customerByJava = TxtConfigurationFactory.getInstanceByJavaReflection("customer_config").asInstanceOf[Customer]
    val customerByScala = TxtConfigurationFactory.getInstanceByScalaReflection[Customer]("customer_config")
    
    /*	Output the information of instances	 */
    println(s"The clerk information:\n${clerkByJava}")
    println(s"The clerk information:\n${clerkByScala}")
    println(s"The customer information:\n${customerByJava}")
    println(s"The customer information:\n${customerByScala}")
  }
  
  /**
   * 	This is a method to test factory method
   */
  private def testFactoryMethod(): Unit = {
    
		/*	Get 2 instances form TWO CHILD configuration factories with the SAME factory reference "ConfigurationFactory"	*/
		var configurationFactory: ConfigurationFactory = new PropConfigurationFactory()
		var lieutenantByJava = configurationFactory.getInstanceByJavaReflection("lieutenant_config").asInstanceOf[Lieutenant]
		println("The lieutenant information: " + lieutenantByJava)
		var lieutenantByScala = configurationFactory.getInstanceByScalaReflection[Lieutenant]("lieutenant_config")
		println("The lieutenant information: " + lieutenantByScala)
		configurationFactory = new XmlConfigurationFactory()
		lieutenantByJava = configurationFactory.getInstanceByJavaReflection("lieutenant_config").asInstanceOf[Lieutenant]
		println("The lieutenant information: " + lieutenantByJava)
		lieutenantByScala = configurationFactory.getInstanceByScalaReflection[Lieutenant]("lieutenant_config")
		println("The lieutenant information: " + lieutenantByScala)
		
		/*	Get another 2 instances form TWO CHILD configuration factories with the SAME factory reference "ConfigurationFactory"	*/
		configurationFactory = new PropConfigurationFactory()
		var sergeantByJava = configurationFactory.getInstanceByJavaReflection("sergeant_config").asInstanceOf[Sergeant]
		println("The sergeant information: " + sergeantByJava)
		var sergeantByScala = configurationFactory.getInstanceByScalaReflection[Sergeant]("sergeant_config")
		println("The sergenat information: " + sergeantByScala)
		configurationFactory = new XmlConfigurationFactory()
		sergeantByJava = configurationFactory.getInstanceByJavaReflection("sergeant_config").asInstanceOf[Sergeant]
		println("The sergeant information: " + sergeantByJava)
		sergeantByScala = configurationFactory.getInstanceByScalaReflection[Sergeant]("sergeant_config")
		println("The sergeant information: " + sergeantByScala)
  }
  
  /**
   * 	This is a method to test abstract factory
   */
  private def testAbstractFactory(): Unit = {
    
		/*	Get 4 instances that belong to TWO families (Mac/Windows + border/button) with the SAME factory reference "BrandFactory"	*/		
		var brandFactory: BrandFactory = new MacFactory()
		val macBorder = brandFactory.getBorder()
		val macButton = brandFactory.getButton()
		brandFactory = new WindowsFactory()
		val windowsBorder = brandFactory.getBorder()
		val windowsButton = brandFactory.getButton()
		macBorder.scroll()
		macButton.design()
		windowsBorder.scroll()
		windowsButton.design()
  }
}

File: JavaAndScala/src/main/java/com/java/se/conclusion/design/factory/TestFactory.java
package com.java.se.conclusion.design.factory;

import com.java.se.conclusion.design.factory.abstracts.BrandFactory;
import com.java.se.conclusion.design.factory.abstracts.MacFactory;
import com.java.se.conclusion.design.factory.abstracts.WindowsFactory;
import com.java.se.conclusion.design.factory.abstracts.common.Border;
import com.java.se.conclusion.design.factory.abstracts.common.Button;
import com.java.se.conclusion.design.factory.configuration.TxtConfigurationFactory;
import com.java.se.conclusion.design.factory.configuration.common.Clerk;
import com.java.se.conclusion.design.factory.configuration.common.Customer;
import com.java.se.conclusion.design.factory.method.ConfigurationFactory;
import com.java.se.conclusion.design.factory.method.PropConfigurationFactory;
import com.java.se.conclusion.design.factory.method.XmlConfigurationFactory;
import com.java.se.conclusion.design.factory.method.common.Lieutenant;
import com.java.se.conclusion.design.factory.method.common.Sergeant;
import com.java.se.conclusion.design.factory.simple.LaptopFactory;
import com.java.se.conclusion.design.factory.simple.VehicleFactory;
import com.java.se.conclusion.design.factory.simple.common.laptop.Laptop;
import com.java.se.conclusion.design.factory.simple.common.vehicle.Vehicle;
import com.java.se.conclusion.design.factory.singleton.LazySingletonFactory;
import com.java.se.conclusion.design.factory.singleton.common.Teacher;

/**
 * 	This is a class to test factory pattern
 * 
 * 	There are several types of factory pattern, of which those commonly-used including:
 *  -- Product-level: Factory for the product
 *     -- Simple factory: to return an instance from different class according to different arguments
 *     -- Singleton factory: to return a singleton instance of any class
 *     -- Configuration factory: to return an instance from configuration file
 *        -- Typically applied in the design of Java frameworks (e.g., Spring, etc.)
 *  -- Factory-level: Based on low-level with factory abstraction
 *     -- Factory method: abstract the factory and delegates the instance creation to its child classes
 *        -- The instance (or product created by factory) belongs to one family
 *     -- Abstract factory: based on factory method and return a multiple families of instance
 *        -- The instance (or product created by factory) belongs to multiple families
 *  
 *  Factory pattern is usually implemented with reflection
 * 
 * @author VinceYuan
 *
 */
public class TestFactory {

	/**
	 * 	This is a main method for execution
	 * 
	 * @param args
	 */
	public static void main(String[] args) {
		TestFactory testFactory = new TestFactory();
		System.out.println("Here tests simple factory:");
		testFactory.testSimpleFactory();
		System.out.println("\nHere tests singleton factory:");
		testFactory.testSingletonFactory();
		System.out.println("\nHere tests configuration factory:");
		testFactory.testConfigurationFactory();
		System.out.println("\nHere tests factory method:");
		testFactory.testFactoryMethod();
		System.out.println("\nHere tests abstract factory:");
		testFactory.testAbstractFactory();
	}
	
	/**
	 * 	Test simple factory pattern
	 */
	private void testSimpleFactory() {
		
		/*	Get different instances of the parent interface according to the argument	*/
		Laptop acer = LaptopFactory.getLaptop("Acer");
		Laptop dell = LaptopFactory.getLaptop("Dell");
		Laptop lenovo = LaptopFactory.getLaptop("Lenovo");
		acer.startup();
		dell.startup();
		lenovo.startup();
		
		/*	Get different instances of the parent abstract class according to the argument	*/
		Vehicle ship = VehicleFactory.getVehicle("Ship");
		Vehicle train = VehicleFactory.getVehicle("Train");
		Vehicle truck = VehicleFactory.getVehicle("Truck");
		ship.turn();
		train.turn();
		truck.turn();
	}
	
	/**
	 * 	Test singleton factory pattern
	 *  -- Singleton factory pattern is collaborated with lazy singleton pattern and double-checked locking pattern
	 */
	private void testSingletonFactory() {
		
		/*	Get 2 instances	 */
		Teacher vince = (Teacher) LazySingletonFactory.getInstance(Teacher.class);
		Teacher violet = (Teacher) LazySingletonFactory.getInstance(Teacher.class);
		
		/*	See if 2 instances are the same	 */
		System.out.println("If vince and violet reference the same instance? " + ((vince == violet) && (vince.equals(violet))));
	}
	
	/**
	 * 	Test configuration factory pattern
	 */
	private void testConfigurationFactory() {
		
		/*	Get 2 different instances from text configuration factory	*/
		Clerk clerk = (Clerk) TxtConfigurationFactory.getInstance("clerk_config");
		Customer customer = (Customer) TxtConfigurationFactory.getInstance("customer_config");
		
		/*	Output the information of the instances	 */
		System.out.println("The clerk information: " + clerk);
		System.out.println("The customer infomration: " + customer);
	}
	
	/**
	 * 	Test factory method pattern
	 */
	private void testFactoryMethod() {
		
		/*	Get 2 instances form TWO CHILD configuration factories with the SAME factory reference "ConfigurationFactory"	*/
		ConfigurationFactory configurationFactory = new PropConfigurationFactory();
		Lieutenant lieutenant = (Lieutenant) configurationFactory.getInstance("lieutenant_config");
		System.out.println("The lieutenant information: " + lieutenant);
		configurationFactory = new XmlConfigurationFactory();
		lieutenant = (Lieutenant) configurationFactory.getInstance("lieutenant_config");
		System.out.println("The lieutenant information: " + lieutenant);
		
		/*	Get another 2 instances form TWO CHILD configuration factories with the SAME factory reference "ConfigurationFactory"	*/
		configurationFactory = new PropConfigurationFactory();
		Sergeant sergeant = (Sergeant) configurationFactory.getInstance("sergeant_config");
		System.out.println("The sergeant information: " + sergeant);
		configurationFactory = new XmlConfigurationFactory();
		sergeant = (Sergeant) configurationFactory.getInstance("sergeant_config");
		System.out.println("The sergeant information: " + sergeant);
	}
	
	/**
	 * 	Test abstract factory pattern
	 */
	private void testAbstractFactory() {
		
		/*	Get 4 instances that belong to TWO families (Mac/Windows + border/button) with the SAME factory reference "BrandFactory"	*/
		BrandFactory brandFactory = new MacFactory();
		Border macBorder = brandFactory.getBorder();
		Button macButton = brandFactory.getButton();
		brandFactory = new WindowsFactory();
		Border windowsBorder = brandFactory.getBorder();
		Button windowsButton = brandFactory.getButton();
		macBorder.scroll();
		macButton.design();
		windowsBorder.scroll();
		windowsButton.design();
	}
}
